### DISCUSSION TITLE
Automatic fixing of errors?

### DISCUSSION THREAD

**Discussion by reutsharabani** [clj-kondo/clj-kondo#1673] [Ideas]
To upvote this issue, give it a thumbs up. See [this list](https://github.com/clj-kondo/clj-kondo/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc) for the most upvoted issues.

**Is your feature request related to a problem? Please describe.**
Some of the errors/warning that come up during lint are trivial but fixing them, especially on newly-linted projects or when adding new linters, is tedious.

**Describe the solution you'd like**
Maybe clj-kondo can supply a --fix flag that can rewrite code as recommended (where it makes sense). An example would be sorting imports. This exists in linters like eslint and rubocop.

**Describe alternatives you've considered**
Currently I am fixing manually, even trivial stuff. For my specific linter I've outputted as a warning the required form to solve the problem (which is deterministic).

**Comment by borkdude** [comment]
@reutsharabani clj-kondo works closely with the clojure-lsp project (of which @ericdallo is currently the most active maintainer) and clojure-lsp provides fixes/refactorings based on what clj-kondo finds. In short, I recommend using the clojure-lsp CLI for this.

**Comment by vincentjames501** [comment]
I'd personally love this feature. Being able to do this directly through a CLI at git commit/push time would be amazing. I don't see an easy way to do this with clojure-lsp. We just added clj-kondo to a bunch of our projects and this would've saved us a ton of time. We had a ton of:

- redundant do/let
- unused imports
- misplaced docstrings
- use :else as catch-all
- use seq
- :refer :all warnings

etc which all seem like they could be autofixed and I suspect this would dramatically improve developer adoption of tools like clj-kondo? Anyways @borkdude would love if this was reconsidered!

**Comment by borkdude** [comment]
@vincentjames501 As said, clojure-lsp already has a lot of that and they offer a CLI to fix those things. I recommend trying it. /cc @ericdallo

**Comment by ericdallo** [comment]
Not sure all of that would be easy to make it work on clojure-lsp, first we would need to have the feature of fix each one of these findings to them provide them via CLI, I think we could start with a issue to discuss a little bit more

**Comment by vincentjames501** [comment]
ESLint has the following which I really like from a CLI usage standpoint FWIW:

```
Fixing problems:
  --fix                           Automatically fix problems
  --fix-dry-run                   Automatically fix problems without saving the changes to the file system
  --fix-type Array                Specify the types of fixes to apply (directive, problem, suggestion, layout)
```

Rust has things like clippy to auto fix things, ruby has rubocop to auto fix issues, etc which are all amazing CLIs.

**Reply by ericdallo** [reply]
clojure-lsp has --dry as well, although I know you would like this on kondo itself

**Comment by borkdude** [comment]
Aside from clojure-lsp there are a few projects that can automatically fix clj-kondo findings:

- https://github.com/borkdude/carve for cutting away unused vars
- https://github.com/oxalorg/clj-konmari/ for auto-configuring inconsistent aliases

**Comment by borkdude** [comment]
My experience is that once you have fixed most initial things on a clojure project, you don't need that much "auto-fixing" anymore since you simply won't introduce any new clj-kondo warnings/errors. My most used "fix" is lsp-organize-imports of clojure-sort-ns (clojure-mode in emacs) for cleaning up an ns form. Other than that, none that I can think of.
